File: src/orchestrix/rate_limit/lua/token_bucket.lua
-- KEYS
local bucket_key = KEYS[1]

-- ARGV
local current_time = tonumber(ARGV[1])
local max_tokens = tonumber(ARGV[2])
local refill_rate = tonumber(ARGV[3])
local tokens_requested = tonumber(ARGV[4])

-- FETCH CURRENT BUCKET STATE
local bucket = redis.call(
    "HMGET",
    bucket_key,
    "tokens",
    "last_refill"
)

local tokens = tonumber(bucket[1])
local last_refill = tonumber(bucket[2])


-- INITIALIZE DEFAULTS
if tokens == nil then
    tokens = max_tokens
end

if last_refill == nil then
    last_refill = current_time
end


-- REFILL LOGIC
local elapsed_time = current_time - last_refill

local tokens_to_add = elapsed_time * refill_rate

local new_tokens = math.min(
    tokens + tokens_to_add,
    max_tokens
)


-- ALLOW / DENY
local allowed = 0
local remaining_tokens = new_tokens

if new_tokens >= tokens_requested then
    remaining_tokens = new_tokens - tokens_requested
    allowed = 1
end


-- RETRY HINT (deterministic; jitter belongs in application code)
local retry_after = 0
if allowed == 0 then
    if refill_rate > 0 then
        retry_after = (tokens_requested - new_tokens) / refill_rate
    else
        retry_after = -1
    end
end


-- PERSIST UPDATED STATE
redis.call(
    "HSET",
    bucket_key,
    "tokens",
    remaining_tokens,
    "last_refill",
    current_time
)

-- Prevent inactive buckets from living forever
redis.call(
    "EXPIRE",
    bucket_key,
    3600
)


-- RETURN RESULT
return {
    allowed,
    remaining_tokens,
    retry_after,
}

File: src/orchestrix/rate_limit/token_bucket.py
import time
from typing import TypedDict


class TokenBucketResult(TypedDict):
    allowed: bool
    remaining_tokens: float


class TokenBucket:
    def __init__(
        self,
        max_tokens: float,
        refill_rate: float,
    ) -> None:
        # Maximum burst capacity
        self.max_tokens = max_tokens

        # Tokens regenerated per second
        self.refill_rate = refill_rate

        # Start with a full bucket
        self.tokens = max_tokens

        # Timestamp of last refill calculation
        self.last_refill = time.time()

    def _try_consume(
        self,
        tokens_requested: float,
    ) -> TokenBucketResult:
        if self.tokens >= tokens_requested:
            # Consume tokens
            self.tokens -= tokens_requested

            return {
                "allowed": True,
                "remaining_tokens": self.tokens,
            }

        return {
            "allowed": False,
            "remaining_tokens": self.tokens,
        }

    def allow(
        self,
        tokens_requested: float = 1,
    ) -> TokenBucketResult:
        # Current timestamp
        now = time.time()

        # Time elapsed since last refill
        elapsed_time = now - self.last_refill

        # Calculate regenerated tokens
        tokens_to_add = elapsed_time * self.refill_rate

        # Refill bucket while respecting max capacity
        self.tokens = min(
            self.tokens + tokens_to_add,
            self.max_tokens,
        )

        # Update refill timestamp
        self.last_refill = now

        # Attempt token consumption
        return self._try_consume(tokens_requested)
